core/src/test/scala/com/evolutiongaming/kafka/flow/PartitionFlowSpec.scala

```diff
@@ -322,7 +322,77 @@ class PartitionFlowSpec extends FunSuite {
+  test(
+    "PartitionFlow commits latest consumed offset on release when commitOnRevoke=true, " +
+      "a TimerFlow with flushOnRevoke=true is used, and all keys successfully persisted their state on release"
+  ) {
+
+    // Waiting for 100 events to represents a long-living flow that doesn't unload the state during its lifetime
+    val f = new ConstFixture(waitForN = 100)
+
+    val flow = f.flow.use { flow =>
+      for {
+        _      <- flow(f.records("key1", 100, List("event1", "event2")))
+        _      <- flow(f.records("key2", 102, List("event1", "event2")))
+        _      <- flow(f.records("key3", 104, List("event1", "event2")))
+        offset <- f.pendingOffset.get
+        // No commits expected to be scheduled yet
+        _ <- IO { assert(clue(offset).isEmpty) }
+      } yield ()
+    } >>
+      // Commit to the last consumed offset should be scheduled during the flow release
+      f.pendingOffset.get.map(offset => assertEquals(offset, Some(Offset.unsafe(106))))
+    flow.unsafeRunSync()
+
+  }
+
+  test(
+    "PartitionFlow commits an offset corresponding to the oldest successfully persisted key on release" +
+      "when commitOnRevoke=true, a TimerFlow with flushOnRevoke=true is used," +
+      "and some keys failed to persist their state on release"
+  ) {
+
+    val key1 = "key1"
+    val key2 = "key2"
+    val key3 = "key3"
+
+    // Waiting for 100 events to represents a long-living flow that doesn't unload the state during its lifetime
+    class LocalFixture extends ConstFixture(100) {
+      private val snapshotDatabase = new SnapshotDatabase[IO, String, State] {
+        def get(key: String): IO[Option[(Offset, Int)]] = IO.pure(none)
+        // Fail snapshot persistence for the second key
+        def persist(key: String, snapshot: (Offset, Int)): IO[Unit] =
+          IO.raiseError(new Exception("Test error")).whenA(key == key2)
+        def delete(key: String): IO[Unit] = IO.unit
+      }
+
+      override def flow: Resource[IO, PartitionFlow[IO]] = makeFlow(
+        timerFlowOf = TimerFlowOf.unloadOrphaned(flushOnRevoke = true),
+        persistenceOf =
+          PersistenceOf.snapshotsOnly(keysOf = keysOf, snapshotsOf = SnapshotsOf.backedBy(snapshotDatabase))
+      )
+    }
+
+    val f = new LocalFixture
+
+    val flow = f.flow.use { flow =>
+      for {
+        // 6 events for 2 different keys are consumed
+        _      <- flow(f.records(key1, 100, List("event1", "event2")))
+        _      <- flow(f.records(key2, 102, List("event1", "event2")))
+        _      <- flow(f.records(key3, 104, List("event1", "event2")))
+        offset <- f.pendingOffset.get
+        // No commits expected to be scheduled yet
+        _ <- IO { assert(clue(offset).isEmpty) }
+      } yield ()
+    } >>
+      // Commit to the first offset corresponding to key2 should be scheduled during the flow release
+      f.pendingOffset.get.map(offset => assertEquals(offset, Some(Offset.unsafe(102))))
```

Comment:
if i understand it correctly, this might be flaky in theory as it relies on an assumption that timer flow for `key1` will be canceled (and thus flushed) before `key2`, for which we don't have any guarantees, right? i don't think `Cache#clear` specifies any ordering guarantees

also, since we're calling `cache.clear` directly in the release function of `PartitionFlow`, how does a failure of one of the keys not get propagated there thus preventing releasing from progressing further? 🤔

Reply:
We don't need any ordering guarantees here because `cache.clear.flatten` will finish when _all_ keys are released.
`flushOnCancel` is resource release nested in the release of a `TimerFlowOf`, which is nested in the release of `KeyFlow`, which is being released when a cache entry is evicted.

Failure to  release any of the keys (which is being tested here) doesn't affect successful release of the other keys.